### PR TITLE
New Resource: aws_memorydb_subnet_group

### PR DESCRIPTION
**Pull request by roberth-k**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates https://github.com/hashicorp/terraform-provider-aws/issues/20631

Output from acceptance testing:

```
$ make testacc PKG=memorydb TESTS=TestAccSubnetGroup_ ACCTEST_PARALLELISM=2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/memorydb/... -v -count 1 -parallel 2 -run='TestAccSubnetGroup_' -timeout 180m
=== RUN   TestAccSubnetGroup_basic
=== PAUSE TestAccSubnetGroup_basic
=== RUN   TestAccSubnetGroup_disappears
=== PAUSE TestAccSubnetGroup_disappears
=== RUN   TestAccSubnetGroup_nameGenerated
=== PAUSE TestAccSubnetGroup_nameGenerated
=== RUN   TestAccSubnetGroup_namePrefix
=== PAUSE TestAccSubnetGroup_namePrefix
=== RUN   TestAccSubnetGroup_update_description
=== PAUSE TestAccSubnetGroup_update_description
=== RUN   TestAccSubnetGroup_update_subnetIds
=== PAUSE TestAccSubnetGroup_update_subnetIds
=== RUN   TestAccSubnetGroup_update_tags
=== PAUSE TestAccSubnetGroup_update_tags
=== CONT  TestAccSubnetGroup_basic
=== CONT  TestAccSubnetGroup_update_description
--- PASS: TestAccSubnetGroup_basic (20.25s)
=== CONT  TestAccSubnetGroup_update_tags
--- PASS: TestAccSubnetGroup_update_description (35.42s)
=== CONT  TestAccSubnetGroup_nameGenerated
--- PASS: TestAccSubnetGroup_nameGenerated (16.95s)
=== CONT  TestAccSubnetGroup_namePrefix
--- PASS: TestAccSubnetGroup_namePrefix (16.81s)
=== CONT  TestAccSubnetGroup_disappears
--- PASS: TestAccSubnetGroup_update_tags (63.59s)
=== CONT  TestAccSubnetGroup_update_subnetIds
--- PASS: TestAccSubnetGroup_disappears (16.08s)
--- PASS: TestAccSubnetGroup_update_subnetIds (48.44s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/memorydb   134.220s
```

### CODE REVIEW

File: internal/service/memorydb/subnet_group.go [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+	tags := defaultTagsConfig.MergeTags(tftags.New(d.Get("tags").(map[string]interface{})))
+
+	var name string
+	if v, ok := d.GetOk("name"); ok {
+		name = v.(string)
+	} else if v, ok := d.GetOk("name_prefix"); ok {
+		name = resource.PrefixedUniqueId(v.(string))
+	} else {
+		name = resource.UniqueId()
+	}
```
Comment (ewbankkit) [inline internal/service/memorydb/subnet_group.go:108]: Drive-by review: You can use the code in the internal `create` package for this.
Documented [here](https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/contributing/contribution-checklists.md#adding-resource-name-generation-support).

Comment (roberth-k) [reply internal/service/memorydb/subnet_group.go]: Thank you for the hint -- that's now in place.

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

##### Commercial

```console
% make testacc PKG=memorydb TESTS=TestAccSubnetGroup_ ACCTEST_PARALLELISM=2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/memorydb/... -v -count 1 -parallel 2 -run='TestAccSubnetGroup_' -timeout 180m
=== RUN   TestAccSubnetGroup_basic
=== PAUSE TestAccSubnetGroup_basic
=== RUN   TestAccSubnetGroup_disappears
=== PAUSE TestAccSubnetGroup_disappears
=== RUN   TestAccSubnetGroup_nameGenerated
=== PAUSE TestAccSubnetGroup_nameGenerated
=== RUN   TestAccSubnetGroup_namePrefix
=== PAUSE TestAccSubnetGroup_namePrefix
=== RUN   TestAccSubnetGroup_update_description
=== PAUSE TestAccSubnetGroup_update_description
=== RUN   TestAccSubnetGroup_update_subnetIds
=== PAUSE TestAccSubnetGroup_update_subnetIds
=== RUN   TestAccSubnetGroup_update_tags
=== PAUSE TestAccSubnetGroup_update_tags
=== CONT  TestAccSubnetGroup_basic
=== CONT  TestAccSubnetGroup_update_description
--- PASS: TestAccSubnetGroup_basic (34.73s)
=== CONT  TestAccSubnetGroup_update_tags
--- PASS: TestAccSubnetGroup_update_description (70.24s)
=== CONT  TestAccSubnetGroup_update_subnetIds
--- PASS: TestAccSubnetGroup_update_tags (133.33s)
=== CONT  TestAccSubnetGroup_nameGenerated
--- PASS: TestAccSubnetGroup_update_subnetIds (100.83s)
=== CONT  TestAccSubnetGroup_namePrefix
--- PASS: TestAccSubnetGroup_nameGenerated (34.40s)
=== CONT  TestAccSubnetGroup_disappears
--- PASS: TestAccSubnetGroup_namePrefix (33.41s)
--- PASS: TestAccSubnetGroup_disappears (30.43s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/memorydb	238.533s
```

##### GovCloud

```console
% make testacc PKG=memorydb TESTS=TestAccSubnetGroup_ ACCTEST_PARALLELISM=2                  
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/memorydb/... -v -count 1 -parallel 2 -run='TestAccSubnetGroup_' -timeout 180m
=== RUN   TestAccSubnetGroup_basic
=== PAUSE TestAccSubnetGroup_basic
=== RUN   TestAccSubnetGroup_disappears
=== PAUSE TestAccSubnetGroup_disappears
=== RUN   TestAccSubnetGroup_nameGenerated
=== PAUSE TestAccSubnetGroup_nameGenerated
=== RUN   TestAccSubnetGroup_namePrefix
=== PAUSE TestAccSubnetGroup_namePrefix
=== RUN   TestAccSubnetGroup_update_description
=== PAUSE TestAccSubnetGroup_update_description
=== RUN   TestAccSubnetGroup_update_subnetIds
=== PAUSE TestAccSubnetGroup_update_subnetIds
=== RUN   TestAccSubnetGroup_update_tags
=== PAUSE TestAccSubnetGroup_update_tags
=== CONT  TestAccSubnetGroup_basic
=== CONT  TestAccSubnetGroup_update_description
=== CONT  TestAccSubnetGroup_basic
    subnet_group_test.go:25: MemoryDB is not supported in aws-us-gov partition
--- SKIP: TestAccSubnetGroup_basic (1.53s)
=== CONT  TestAccSubnetGroup_update_description
    subnet_group_test.go:25: MemoryDB is not supported in aws-us-gov partition
--- SKIP: TestAccSubnetGroup_update_description (1.53s)
=== CONT  TestAccSubnetGroup_namePrefix
=== CONT  TestAccSubnetGroup_nameGenerated
=== CONT  TestAccSubnetGroup_namePrefix
    subnet_group_test.go:25: MemoryDB is not supported in aws-us-gov partition
--- SKIP: TestAccSubnetGroup_namePrefix (0.00s)
=== CONT  TestAccSubnetGroup_update_tags
=== CONT  TestAccSubnetGroup_nameGenerated
    subnet_group_test.go:25: MemoryDB is not supported in aws-us-gov partition
--- SKIP: TestAccSubnetGroup_nameGenerated (0.00s)
=== CONT  TestAccSubnetGroup_disappears
=== CONT  TestAccSubnetGroup_update_tags
    subnet_group_test.go:25: MemoryDB is not supported in aws-us-gov partition
=== CONT  TestAccSubnetGroup_disappears
    subnet_group_test.go:25: MemoryDB is not supported in aws-us-gov partition
--- SKIP: TestAccSubnetGroup_update_tags (0.00s)
=== CONT  TestAccSubnetGroup_update_subnetIds
--- SKIP: TestAccSubnetGroup_disappears (0.00s)
=== CONT  TestAccSubnetGroup_update_subnetIds
    subnet_group_test.go:25: MemoryDB is not supported in aws-us-gov partition
--- SKIP: TestAccSubnetGroup_update_subnetIds (0.00s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/memorydb	6.430s
```

**Comment by ewbankkit** [comment]
@roberth-k Thanks for the contribution :tada: :clap:.
